src/contract/assembled_transaction.ts

```diff
@@ -411,21 +429,70 @@ export class AssembledTransaction<T> {
     return tx;
   }
 
-  simulate = async (): Promise<this> => {
+  private static async buildFootprintRestoreTransaction<T>(
+    options: AssembledTransactionOptions<T>,
+    sorobanData: SorobanDataBuilder | xdr.SorobanTransactionData,
+    account: Account,
+    fee: string
+  ): Promise<AssembledTransaction<T>> {
+    const tx = new AssembledTransaction(options);
+    tx.raw = new TransactionBuilder(account, {
+      fee,
+      networkPassphrase: options.networkPassphrase,
+    })
+      .setSorobanData(sorobanData instanceof SorobanDataBuilder ? sorobanData.build() : sorobanData)
+      .addOperation(Operation.restoreFootprint({}))
+      .setTimeout(options.timeoutInSeconds ?? DEFAULT_TIMEOUT);
+    await tx.simulate(false);
+    return tx;
+  }
+
+  simulate = async (restore?: boolean): Promise<this> => {
     if (!this.raw) {
       throw new Error(
         "Transaction has not yet been assembled; " +
-        "call `AssembledTransaction.build` first.",
+        "call `AssembledTransaction.build` first."
       );
     }
 
+    restore = restore ?? this.options.restore;
     this.built = this.raw.build();
     this.simulation = await this.server.simulateTransaction(this.built);
 
+    if (Api.isSimulationRestore(this.simulation) && restore) {
+      const account = await AssembledTransaction.getAccount(this.options, this.server);
+      const result = await this.restoreFootprint(
+        this.simulation.restorePreamble,
+        account
+      );
+      if (result.status === Api.GetTransactionStatus.SUCCESS) {
+        // need to rebuild the transaction with bumped account sequence number
+        const contract = new Contract(this.options.contractId);
+        this.raw = new TransactionBuilder(account, {
+          fee: this.options.fee ?? BASE_FEE,
+          networkPassphrase: this.options.networkPassphrase,
+        })
+          .addOperation(
+            contract.call(
+              this.options.method,
+              ...(this.options.args ?? [])
+            )
+          )
+          .setTimeout(
+            this.options.timeoutInSeconds ?? DEFAULT_TIMEOUT
+          );
```

Comment:
This is exactly the same as the logic inside `AssembledTransaction.build`. Should we extract it to a common location that both can use? (If the answer is "no", that's fine! But tell me why.)

Reply:
I think that would express our logic here _much_ more clearly. Have you explored LoE on that? js-stellar-base is not super fun to work with, but these sound like small changes. @Shaptic do you foresee challenges to this approach?

Reply:
Another option: let's go ahead and ship this as-is, then do the sequence-number-tweaking in stellar-base as a follow up, then circle-back and clean up the logic here in the spots where it's just working around the current sequence-number-tweaking limitations.

Reply:
I think as-is then refactor will let us avoid version fragmentation since I need to hurry and tag off stable releases.

Reply:
Sounds good, I'll keep this logic for now.